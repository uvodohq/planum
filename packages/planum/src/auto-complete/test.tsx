import {
  autoUpdate,
  size,
  useDismiss,
  useFloating,
  useId,
  useInteractions,
  useListNavigation,
  useRole,
} from '@floating-ui/react-dom-interactions'
import React, { forwardRef, useLayoutEffect, useRef, useState } from 'react'

interface ItemProps {
  children: React.ReactNode
  active: boolean
}

const Item = forwardRef<
  HTMLDivElement,
  ItemProps & React.HTMLProps<HTMLDivElement>
>(({ children, active, ...rest }, ref) => {
  const id = useId()
  return (
    <div
      ref={ref}
      role="option"
      id={id}
      aria-selected={active}
      {...rest}
      style={{
        background: active ? 'lightblue' : 'none',
        padding: 4,
        cursor: 'default',
        ...rest.style,
      }}>
      {children}
    </div>
  )
})
Item.displayName = 'Item'

export function AutoCompleteTest({ data }) {
  const [open, setOpen] = useState(false)
  const [inputValue, setInputValue] = useState('')
  const [activeIndex, setActiveIndex] = useState<number | null>(null)

  const listRef = useRef<Array<HTMLElement | null>>([])

  const { x, y, reference, floating, strategy, context, refs } =
    useFloating<HTMLInputElement>({
      whileElementsMounted: autoUpdate,
      open,
      onOpenChange: setOpen,
      middleware: [
        size({
          apply({ rects, elements }) {
            Object.assign(elements.floating.style, {
              width: `${rects.reference.width}px`,
              maxHeight: `${200}px`,
            })
          },
          padding: 10,
        }),
      ],
    })

  useLayoutEffect(() => {
    // IMPORTANT: When the floating element first opens, this runs when the
    // styles have **not yet** been applied to the element. This can cause an
    // infinite loop as `size` has not yet limited the maxHeight, so the whole
    // page tries to scroll. We must wrap it in rAF.
    requestAnimationFrame(() => {
      if (activeIndex != null) {
        listRef.current[activeIndex]?.scrollIntoView({ block: 'nearest' })
      }
    })
  }, [activeIndex])

  const { getReferenceProps, getFloatingProps, getItemProps } = useInteractions(
    [
      useRole(context, { role: 'listbox' }),
      useDismiss(context),
      useListNavigation(context, {
        listRef,
        activeIndex,
        onNavigate: setActiveIndex,
        virtual: true,
        loop: true,
        focusItemOnHover: false,
      }),
    ],
  )

  function onChange(event: React.ChangeEvent<HTMLInputElement>) {
    const value = event.target.value
    setInputValue(value)

    if (value) {
      setOpen(true)
      setActiveIndex(0)
    } else {
      setOpen(false)
    }
  }

  const items = data.filter((item) =>
    item.toLowerCase().startsWith(inputValue.toLowerCase()),
  )

  return (
    <>
      <input
        {...getReferenceProps({
          ref: reference,
          onChange,
          value: inputValue,
          placeholder: 'Enter fruit',
          'aria-autocomplete': 'list',
          onKeyDown(event) {
            if (
              event.key === 'Enter' &&
              activeIndex != null &&
              items[activeIndex]
            ) {
              setInputValue(items[activeIndex])
              setActiveIndex(null)
              setOpen(false)
            }

            if (event.key === 'Tab') {
              setOpen(false)
            }
          },
        })}
      />
      {open && (
        <div
          {...getFloatingProps({
            ref: floating,
            style: {
              position: strategy,
              left: x ?? 0,
              top: y ?? 0,
              background: '#eee',
              color: 'black',
              overflowY: 'auto',
            },
          })}>
          {items.map((item, index) => (
            // eslint-disable-next-line react/jsx-key
            <Item
              {...getItemProps({
                key: item,
                ref(node) {
                  listRef.current[index] = node
                },
                onClick() {
                  setInputValue(item)
                  setOpen(false)
                  refs.reference.current?.focus()
                },
              })}
              active={activeIndex === index}>
              {item}
            </Item>
          ))}
        </div>
      )}
    </>
  )
}
