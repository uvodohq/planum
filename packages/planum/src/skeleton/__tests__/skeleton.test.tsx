import { vi } from 'vitest'

import { render, screen } from '~/test'

import { Skeleton } from '..'

describe('<Skeleton />', async () => {
  const originalRandom = Math.random

  beforeEach(() => {
    const mocked = vi.fn(() => 0.3)
    Math.random = mocked
  })

  afterEach(() => {
    Math.random = originalRandom
  })

  it('should render default skeleton as 2 line texts', () => {
    render(<Skeleton />)

    expect(screen.getByRole('progressbar')).toMatchSnapshot()
  })

  it('should render default skeleton as item card with image', () => {
    render(<Skeleton image />)

    screen.getByTestId('skeleton-img')

    expect(screen.getByRole('progressbar')).toMatchSnapshot()
  })

  it('should render default skeleton as Single image', () => {
    render(<Skeleton showText={false} image />)

    expect(screen.getByRole('progressbar')).toMatchSnapshot()
  })

  it('should render default skeleton as 1 line paragraph', () => {
    render(<Skeleton oneTextLine />)

    expect(screen.getByRole('progressbar')).toMatchSnapshot()
  })
})
