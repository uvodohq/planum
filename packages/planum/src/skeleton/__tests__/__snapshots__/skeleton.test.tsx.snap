// Vitest Snapshot v1

exports[`<Skeleton /> > should render default skeleton as 1 line paragraph 1`] = `
<div
  aria-busy="true"
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuetext="loading"
  class="c-lfOXhj c-lfOXhj-hGdOCN-animated-true"
  role="progressbar"
>
  <div
    class="c-cRMbGV"
  >
    <div
      class="c-fCqUlT"
      style="width: 45%;"
    >
      <p
        class="c-iPYhUp c-kLRYwF c-hAVAWC c-iPYhUp-iiODFRR-css"
      />
    </div>
  </div>
</div>
`;

exports[`<Skeleton /> > should render default skeleton as 2 line texts 1`] = `
<div
  aria-busy="true"
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuetext="loading"
  class="c-lfOXhj c-lfOXhj-hGdOCN-animated-true"
  role="progressbar"
>
  <div
    class="c-cRMbGV"
  >
    <div
      class="c-fCqUlT"
      style="width: 45%;"
    >
      <p
        class="c-iPYhUp c-kLRYwF c-hAVAWC c-iPYhUp-iiODFRR-css"
      />
    </div>
    <div
      class="c-fCqUlT"
      style="width: 20%;"
    >
      <p
        class="c-iPYhUp c-kLRYwF c-hAVAWC c-iPYhUp-iiODFRR-css"
      />
    </div>
  </div>
</div>
`;

exports[`<Skeleton /> > should render default skeleton as Single image 1`] = `
<div
  aria-busy="true"
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuetext="loading"
  class="c-lfOXhj c-lfOXhj-hGdOCN-animated-true"
  role="progressbar"
>
  <div
    class="c-dbbEBI c-dbbEBI-iiODFRR-css"
    data-testid="skeleton-img"
  />
</div>
`;

exports[`<Skeleton /> > should render default skeleton as item card with image 1`] = `
<div
  aria-busy="true"
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuetext="loading"
  class="c-lfOXhj c-lfOXhj-hGdOCN-animated-true"
  role="progressbar"
>
  <div
    class="c-dbbEBI c-dbbEBI-iiODFRR-css"
    data-testid="skeleton-img"
  />
  <div
    class="c-cRMbGV"
  >
    <div
      class="c-fCqUlT"
      style="width: 45%;"
    >
      <p
        class="c-iPYhUp c-kLRYwF c-hAVAWC c-iPYhUp-iiODFRR-css"
      />
    </div>
    <div
      class="c-fCqUlT"
      style="width: 20%;"
    >
      <p
        class="c-iPYhUp c-kLRYwF c-hAVAWC c-iPYhUp-iiODFRR-css"
      />
    </div>
  </div>
</div>
`;
